src/core/credrank/personalAttribution.js

```diff
@@ -0,0 +1,150 @@
+// @flow
+
+import findLast from "lodash.findlast";
+import {type IdentityId, identityIdParser} from "../identity/id";
+import type {TimestampMs} from "../../util/timestamp";
+import * as C from "../../util/combo";
+
+export type PersonalAttributionProportion = {|
+  +timestampMs: TimestampMs,
+  +proportionValue: number,
+|};
+export type PersonalAttribution = {|
+  +fromParticipantId: IdentityId,
+  +toParticipantId: IdentityId,
+  +proportions: $ReadOnlyArray<PersonalAttributionProportion>,
+|};
+export type PersonalAttributions = $ReadOnlyArray<PersonalAttribution>;
+export const personalAttributionsParser: C.Parser<PersonalAttributions> = C.array(
+  C.object({
+    fromParticipantId: identityIdParser,
+    toParticipantId: identityIdParser,
+    proportions: C.array(
+      C.object({timestampMs: C.number, proportionValue: C.number})
+    ),
+  })
+);
+
+type GroupedPersonalAttributions = PersonalAttributions;
+export class PersonalAttributionsMap {
+  _map: $ReadOnlyMap<IdentityId, GroupedPersonalAttributions>;
+
+  constructor(
+    personalAttributions: PersonalAttributions,
+    epochStarts: $ReadOnlyArray<TimestampMs>
+  ) {
+    // Validate that:
+    // 1. There is only 1 entry per [fromParticipantId, toParticipantId] pair.
+    // 2. Proportions are in chronological order.
+    // 3. Proportions are a number between 0 and 1.
+    const identityPairs = new Set();
+    for (const pa of personalAttributions) {
+      if (identityPairs.has([pa.fromParticipantId, pa.toParticipantId]))
+        throw `More than one attribution found from [${pa.fromParticipantId}] to [${pa.toParticipantId}]`;
```

Comment:
Hmm I think it's really important that we allow this value to change over time--e.g. Garnet gives Amethyst 20% in 2020 but only 10% in 2021 -- as I understand the type, that would be implemented by having multiple entries with teh same from->to but different proportions and timestamps.